pandas/core/arrays/_reshaping.py

```diff
@@ -0,0 +1,91 @@
+"""
+Utilities for implementing 2D compatibility for 1D ExtensionArrays.
+"""
+from typing import Tuple
+
+import numpy as np
+
+from pandas._libs.lib import is_integer
+
+
+def implement_2d(cls):
```

Comment:
what is the reasoning behind making this a decorator as opposed to just having base class method? it seems much simpler and more obvious.

Reply:
The main pushback on this idea was the onus put on EA authors who only support 1D.  ATM this decorator mainly just renames `__len__` to `size` so that it can re-define `shape` in a 2D-compatible way.  So asking authors to use the decorator instead of doing that re-definition themselves is kind of a wash.

But the step after this is to patch `__getitem__`, `__setitem__`, `take`, `__iter__`, all of which are easier to do with the decorator than by asking authors to do it themselves.

Reply:
Is the plan still to have an "opt me out of this, I natively support 2D arrays"? *If* we go down the route of putting every Block.values inside an EA (with PandasArray for the current NumPy-backed Blocks), then we'll want that, right?

Reply:
Oh, I see you're applying this to subclasses, rather than ExtensionArray itself. Motivation for that?

Reply:
> Motivation for that?

less magic than a metaclass (and my first attempt at using a metaclass failed)

> Is the plan still to have an "opt me out of this, I natively support 2D arrays"? 

Defined `EA._allows_2d = False`.  Authors set that to True if they implement this themselves.  This decorator should be updated to check that and be a no-op in that case.

Reply:
Oh, right, because simply decorating

```python
@implements_2d
class ExtensionArray:
```

would work for subclasses right? It has to be a metaclass. I'll poke around at the meta class approach to see what's going on. May be too magical.

But if we do go with a